src/server/game/Loot/LootItemStorage.cpp

```diff
@@ -0,0 +1,364 @@
+    stmt->setUInt32(0, containerId);
+    trans->Append(stmt);
+
+    stmt = CharacterDatabase.GetPreparedStatement(CHAR_DEL_ITEMCONTAINER_MONEY);
+    stmt->setUInt32(0, containerId);
+    trans->Append(stmt);
+
+    CharacterDatabase.CommitTransaction(trans);
+}
+
+void LootItemStorage::RemoveStoredLootItemForContainer(uint32 containerId, uint32 itemId, uint32 count)
+{
+    // write
+    boost::unique_lock<boost::shared_mutex> lock(*GetLock());
+
+    auto itr = _lootItemStore.find(containerId);
+    if (itr == _lootItemStore.end())
+        return;
+
+    itr->second.RemoveItem(itemId, count);
+}
+
+void LootItemStorage::AddNewStoredLoot(Loot* loot, Player* player)
+{
+    // Saves the money and item loot associated with an openable item to the DB
+    if (loot->isLooted()) // no money and no loot
+        return;
+
+    // read
+    {
+        boost::shared_lock<boost::shared_mutex> lock(*GetLock());
+
+        auto itr = _lootItemStore.find(loot->containerID);
+        if (itr != _lootItemStore.end())
+        {
+            TC_LOG_ERROR("misc", "Trying to store item loot by player: %s for container id: %u that is already in storage!", player->GetGUID().ToString().c_str(), loot->containerID);
+            return;
+        }
+    }
+
+    StoredLootContainer container;
+    container.SetContainer(loot->containerID);
+
+    SQLTransaction trans = CharacterDatabase.BeginTransaction();
+    if (loot->gold)
+        container.AddMoney(loot->gold, trans);
+
+    PreparedStatement* stmt = CharacterDatabase.GetPreparedStatement(CHAR_DEL_ITEMCONTAINER_ITEMS);
+    stmt->setUInt32(0, loot->containerID);
+    trans->Append(stmt);
+
+    for (LootItem const& li : loot->items)
+    {
+        // Conditions are not checked when loot is generated, it is checked when loot is sent to a player.
+        // For items that are lootable, loot is saved to the DB immediately, that means that loot can be
+        // saved to the DB that the player never should have gotten. This check prevents that, so that only
+        // items that the player should get in loot are in the DB.
+        // IE: Horde items are not saved to the DB for Ally players.
+        if (!li.AllowedForPlayer(player))
+            continue;
+
+        // Don't save currency tokens
+        ItemTemplate const* itemTemplate = sObjectMgr->GetItemTemplate(li.itemid);
+        if (!itemTemplate || itemTemplate->IsCurrencyToken())
+            continue;
+
+        container.AddLootItem(li, trans);
+    }
+
+    CharacterDatabase.CommitTransaction(trans);
+
+    // write
+    {
+        boost::unique_lock<boost::shared_mutex> lock(*GetLock());
+        _lootItemStore.emplace(loot->containerID, std::move(container));
+    }
+}
+
+void StoredLootContainer::SetContainer(uint32 containerId)
+{
+    _containerId = containerId;
+}
+
+void StoredLootContainer::AddLootItem(LootItem const& lootItem, SQLTransaction& trans)
+{
+    _lootItems.emplace(lootItem.itemid, lootItem);
```

Comment:
Travis doesn't like this. Any ideas? it compiles just fine on a VM with Clang 3.9 under Ubuntu Server 16.10